backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -160,15 +186,14 @@ public String getNeededData() {
         String needDatum = null;
         String nextKey = null;
         for (Entry e : possibleEntries) {
-
             SessionDatum datumNeededForThisEntry = getFirstMissingDatum(this.getData(), e.getSessionDataReqs());
             if (datumNeededForThisEntry != null) {
                 String needed = datumNeededForThisEntry.getDataId();
                 if (nextKey == null) {
                     nextKey = needed;
-                    if (datumNeededForThisEntry.getNodeset() != null) {
+                    if (datumNeededForThisEntry instanceof EntityDatum) {
                         needDatum = SessionFrame.STATE_DATUM_VAL;
-                    } else {
+                    } else if (datumNeededForThisEntry instanceof ComputedDatum) {
```

Comment:
Depending on how `FormIdDatum`s are used, adding this `if` to the `else` might be invalid